### ISSUE TITLE
Mongo Repository and optimistic locking 

### ISSUE THREAD

**Issue by asereda-gs** [immutables/immutables#622]
Hello,

Similar to [findOneAndReplace()](https://docs.mongodb.com/manual/reference/method/db.collection.findOneAndReplace/) mongo operation (which was introduced in v3.2) is it currently possible to replace **whole document** in mongo (based on a criteria) ? Like a [CAS](https://en.wikipedia.org/wiki/Compare-and-swap) operation for atomic types.

I'm aware of partial `$set` operations (from Updater) like `set` / `init` / `remove` but none of them provide the option to change full document (not just sub-elements).

### Example
```java
// get latest version from DB
Document doc = repository.findById(id).fetchFirst().getUnchecked().get();
// apply business logic (note version increment)
Document newDoc = businessLogic(doc).withVersion(doc.version() + 1);
// override existing document IF version of stored document matches 
// if not, retry business logic because it was previously applied on a stale data
repository.update(repository.criteria().id(doc.id()).version(doc.version())) // criteria for update
    .replace(newDoc) // this is new replacement
    .updateFirst()
    .getUnchecked();
```

Would you consider such a contribution ?

**Comment by elucash** [comment]
I'm certainly agree of the need of such operation. I somehow thought that that should be already be possible, but after looking into seems that there are no viable alternatives to patch all attributes.
I'm not sure we need `.replace` on Updater, but maybe. But we, probably, need it on Modifier:

```java
items.find(where...)
   .andModifyFirst()
   .replace(newItem) // or alternatively .andReplaceFirst(newItem) instead
```

**Comment by asereda** [comment]
Eugene,

What do you think about introducing `@Mongo.Version` annotation ? Or should user be responsible for 
 specifying / incrementing version number manually in the query ?

**Comment by asereda** [comment]
Also is there any reason why `$set` operation is missing for arrays on Modifier / Updater ?
Sometimes I would like to override whole sub-array not just append / remove an element from it.

Doing `clearFoo().addFoo(foo)` simultaneously does not work.

**Comment by elucash** [comment]
About `$set` I don't know, probably just omission, have to take a look. `clearFoo().addFoo(foo)` will not work as only single operation per attribute will be sent to mongodb.
About `@Mongo.Version` I'm not sure. One part of me says "sure, why not" the other part of me is quite skeptical about it. Here's my thinking:

* If version annotation would serve for automatic version handling:
  + we cannot, probably, guarantee it would work for all operations or combinations of operations, therefore users should be very mindful about what operation to invoke to make sure it will work.
  + auto handling of version fields often give people illusion that they might not worry about properly handling optimistic updated failures and execute proper merge/retry where appropriate.
* If version annotation would serve for generation of conveniences like `.incrementVersion()` or `criteria.havingVersion(v)` then
  + Those conveniences would be of little value compared to just having `int version()` attribute name (with whatever BSON name attached with `SerializedName` annotation) and using generated methods as they are now.

**Comment by asereda** [comment]
I will submit a PR for `$set` on arrays.

Regarding `replace` (`$set` on a whole document). I would like to get your preference (advice). 

Assuming it is part of `Modifier` class do you recommend creating separate Constraint (which seems an overkill) or manually loop through all attributes (currently fieldname is expected for each new constraint) ? 
```java
   public Modifier replace(Foo document) {
      setName(document.name());
      setAge(document.age());
      setList(document.list());
      ...
      return this;
   }
```
Or add replace method as part of Finder (?). 
```java
class Finder {
  /**
    * Don't allow any modifications after replace
   */
  Repositories.Modifier<Foo, Modifier> andReplaceFirst(newItem); // (?)
}
```
In this case, I still need to `$set` whole document (field by field ?) as in previous example.

Thanks for your help.

**Comment by elucash** [comment]
1. It depends how we want it to be on the wire. If we want plain document replacement in BSON rather than series of "$set" then we might use separate constraint with name like `*` and some special handling for it. I just don't remember how it works exactly, so may be wrong

2. If we want to have `andReplaceFirst` we probably need `Replacer` class containing only `upset`, `returning(New|Old)` etc, but not individual properties. If we don't want to create yet another helper, we might just add `replace` to modifier and add runtime checks so attribute modifiers will throw `IllegalStateException` if called after replace or replace is called after modifiers.

**Comment by asereda** [comment]
Probably I'm leaning towards `Replacer`. Just because compile-time checks are, usually, preferable. Also in v3 it can be implemented natively with [findOneAndReplace](https://docs.mongodb.com/manual/reference/method/db.collection.findOneAndReplace/) mongo function.

Let me know if you think otherwise. If not, I'll try to submit a PR.

**Comment by elucash** [comment]
I agree, and the usage of `findOneAndReplace` adds to it as a natural way to model this operation with separate `Replacer`

**Comment by asereda-gs** [comment]
PR #624 for `$set` on arrays

**Comment by asereda-gs** [comment]
PR #626 for full document replace based on criteria

**Comment by asereda-gs** [comment]
Eugene,

Can you make a new release with these changes ? It will help us remove local fork of immutables library.

**Comment by elucash** [comment]
sure, but will do this tomorrow, it takes some time to build and propagate to central. Thank you for PR

**Comment by elucash** [comment]
released 2.5.2
